client/core/core.go

```diff
@@ -5747,6 +5816,109 @@ func (c *Core) authDEX(dc *dexConnection) error {
+	}
+
+	// Identify bonds we consider live that are either pending or missing from
+	// server. In either case, do c.monitorBondConfs (will be immediate postBond
+	// and bondConfirmed if at required confirmations).
+	for _, bond := range localActiveBonds {
+		key := bondKey(bond.AssetID, bond.CoinID)
+		symb := dex.BipIDSymbol(bond.AssetID)
+		bondIDStr := coinIDString(bond.AssetID, bond.CoinID)
+
+		_, found := remoteLiveBonds[key]
+		if found {
+			continue // good, it's live server-side too
+		}
+
+		c.log.Warnf("Locally-active bond %v (%s) not reported by server. Will repost...",
+			bondIDStr, symb) // unexpected, but postbond again
+
+		bondAsset := bondAssets[symb]
+		if bondAsset == nil {
+			c.log.Warnf("Server no longer supports %v as a bond asset!", symb)
+			continue
+		}
+
+		// Unknown on server. postBond at required confs.
+		c.log.Infof("Posting locally-confirmed bond %v (%s).", bondIDStr, symb)
+		c.monitorBondConfs(dc, assetBond(bond), bondAsset.Confs)
+		continue
+	}
+
+	// Identify bonds we consider pending that are either live or missing from
+	// server. If live on server, do c.bondConfirmed. If missing, do
+	// c.monitorBondConfs.
+	for _, bond := range localPendingBonds {
+		key := bondKey(bond.AssetID, bond.CoinID)
+		symb := dex.BipIDSymbol(bond.AssetID)
+		bondIDStr := coinIDString(bond.AssetID, bond.CoinID)
+
+		_, found := remoteLiveBonds[key]
+		if found {
+			// It's live server-side. Confirm it locally.
+			c.log.Debugf("Confirming pending bond %v that is confirmed server side", bondIDStr)
+			if err = c.bondConfirmed(dc, bond.AssetID, bond.CoinID, tier /* no change */); err != nil {
+				c.log.Errorf("Unable to confirm bond %s: %v", bondIDStr, err)
+			}
+			continue
+		}
+
+		c.log.Debugf("Starting coin waiter for pending bond %v (%s)", bondIDStr, symb)
+		bondAsset := bondAssets[symb]
+		if bondAsset == nil {
+			c.log.Warnf("Server no longer supports %v as a bond asset!", symb)
+			continue // will retry, eventually refund
+		}
+
+		// Still pending on server. Start waiting for confs.
+		c.log.Debugf("Preparing to post pending bond %v (%s).", bondIDStr, symb)
+		c.monitorBondConfs(dc, assetBond(bond), bondAsset.Confs)
+	}
+
+	localBondMap := make(map[string]struct{}, len(localActiveBonds)+len(localPendingBonds))
+	for _, dbBond := range localActiveBonds {
+		localBondMap[bondKey(dbBond.AssetID, dbBond.CoinID)] = struct{}{}
+	}
+	for _, dbBond := range localPendingBonds {
+		localBondMap[bondKey(dbBond.AssetID, dbBond.CoinID)] = struct{}{}
+	}
+
+	for _, bond := range result.ActiveBonds {
+		key := bondKey(bond.AssetID, bond.CoinID)
+		if _, found := localBondMap[key]; found {
+			continue
+		}
+		// EXPERIMENT: Server knows of a bond we do not! Store what we can for
+		// tier accounting, but we can't redeem it. We'll make en entry in
+		// dc.acct.bonds for tier accounting, but this may not work out. i.e.
+		// What handling in the bond monitor / refund loop do we need to discard
+		// these after expiry?
```

Comment:
> Shouldn't we make an attempt to find the keys that created this?

We can search our DB, but they should have already been loaded into `localBondMap`, otherwise it would have already been redeemed and we don't need the keys, so this whole case feels like server spamming.

I think I just wanted to have it in our active bonds slice so we can locally match the bond-based tier that the server seems to have.

> If it cannot be found, it could be removed when a tier change note or an expired bond note arrives.

That's a good plan.  Just have the bond monitor / refund loop ignore anything without a PrivKey or a RefundTx, and let them stay until server says they are not longer active.

Reply:
I meant to search using `RefundBondByCoinID` once it exists. But yes I think it's good to have in the active bonds slice.